airbyte-commons/src/main/java/io/airbyte/commons/logging/MaskedDataInterceptor.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.appender.rewrite.RewritePolicy;
+import org.apache.logging.log4j.core.config.plugins.Plugin;
+import org.apache.logging.log4j.core.config.plugins.PluginAttribute;
+import org.apache.logging.log4j.core.config.plugins.PluginFactory;
+import org.apache.logging.log4j.core.impl.Log4jLogEvent;
+import org.apache.logging.log4j.message.SimpleMessage;
+import org.apache.logging.log4j.status.StatusLogger;
+
+/**
+ * Custom Log4j2 {@link RewritePolicy} used to intercept all log messages and mask any JSON
+ * properties in the message that match the list of maskable properties.
+ * <p>
+ * The maskable properties file is generated by a Gradle task in the {@code :airbyte-config:specs}
+ * project. The file is named {@code specs_secrets_mask.yaml} and is located in the
+ * {@code src/main/resources/seed} directory of the {@link :airbyte-config:init} project.
+ */
+@Plugin(name = "MaskedDataInterceptor",
+        category = "Core",
+        elementType = "rewritePolicy",
+        printObject = true)
+public class MaskedDataInterceptor implements RewritePolicy {
+
+  protected static final Logger logger = StatusLogger.getLogger();
+
+  /**
+   * Mask string used as a replacement value for any sensitive properties found in the log message.
+   */
+  public static final String MASK = "********";
+
+  /**
+   * The pattern used to determine if a message contains sensitive data.
+   */
+  private final Optional<String> pattern;
+
+  @PluginFactory
+  public static MaskedDataInterceptor createPolicy(
+                                                   @PluginAttribute(value = "specMaskFile",
+                                                                    defaultString = "/seed/specs_secrets_mask.yaml") final String specMaskFile) {
+    return new MaskedDataInterceptor(specMaskFile);
+  }
+
+  private MaskedDataInterceptor(final String specMaskFile) {
+    this.pattern = buildPattern(specMaskFile);
+  }
+
+  @Override
+  public LogEvent rewrite(final LogEvent source) {
+    return Log4jLogEvent.newBuilder()
+        .setLoggerName(source.getLoggerName())
+        .setMarker(source.getMarker())
+        .setLoggerFqcn(source.getLoggerFqcn())
+        .setLevel(source.getLevel())
+        .setMessage(new SimpleMessage(applyMask(source.getMessage().getFormattedMessage())))
+        .setThrown(source.getThrown())
+        .setContextMap(source.getContextMap())
+        .setContextStack(source.getContextStack())
+        .setThreadName(source.getThreadName())
+        .setSource(source.getSource())
+        .setTimeMillis(source.getTimeMillis())
+        .build();
+  }
+
+  /**
+   * Applies the mask to the message, if necessary.
+   *
+   * @param message The log message.
+   * @return The possibly masked log message.
+   */
+  private String applyMask(final String message) {
+    if (pattern.isPresent()) {
```

Comment:
would it make sense to trigger an alert whenever this happens? this approach is a last-ditch protection, so when it does save us, we should alert ourselves so that we can fix the actual root cause. it has the added benefit of making it so that _we_ are the ones who catch the error. i think we would treat these alerts with high priority.

Reply:
@cgardens What type of alert do you have in mind?  We could certainly put a metric in here and monitor for that to be hit, but that might be a bit hard to tune as there may be legitimate cases where we want to log the object with the secrets masked.  I'm curious about how we could differentiate between accidental use of the mask and legitimate uses of the mask or is that even a concern (e.g. we should never, ever log those objects even for debug purposes).

Reply:
I think it's the latter. We should never, ever log them. This PR is the last line of defense. If we get to here, we goofed upstream and should address upstream ASAP.

The _only_ caveat I could see if for local development (or maybe in a dev cluster). So making env dependent seems reasonable. For any production use case (OSS or Cloud) though, these should never, ever be logged.

Reply:
At the very least we should do a DD counter. That at least covers all of Cloud. If we want to cover OSS too, we would have to do something in Segment. I'm not sure how we would then hook that into our alerting.

Reply:
We don't need to block merging this PR on this btw. This is strictly better than what we have. But I would like us to resolve how we will do alerting on this in the next week or so. Don't want to forget it.